Cannot build on Ubuntu 22 with CUDA 12

### Checklist

- [X] I have searched for [similar issues](https://github.com/isl-org/Open3D/issues).
- [X] For Python issues, I have tested with the [latest development wheel](http://www.open3d.org/docs/latest/getting_started.html#development-version-pip).
- [X] I have checked the [release documentation](http://www.open3d.org/docs/release/) and the [latest documentation](http://www.open3d.org/docs/latest/) (for `master` branch).

### Steps to reproduce the issue

#### I first cloned Open3D by:
```
git clone https://github.com/isl-org/Open3D.git
cd Open3D
```

#### Then, I build Open3D (on Ubuntu 22.04, with CUDA 12.0) with:
```
mkdir build
cd build
cmake -DBUILD_CUDA_MODULE=ON ..
make -j$(nproc)
```


### Error message

```shell
[ 81%] Building CXX object cpp/open3d/CMakeFiles/Open3D.dir/Open3DConfig.cpp.o
[ 81%] Linking CXX static library ../../lib/Release/libOpen3D.a
[ 81%] Built target Open3D
make[2]: *** No rule to make target '/usr/local/cuda/lib64/liblapack_static.a', needed by 'bin/GLInfo'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target '/usr/local/cuda/lib64/liblapack_static.a', needed by 'bin/ViewGeometry'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target '/usr/local/cuda/lib64/liblapack_static.a', needed by 'bin/MergeMesh'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target '/usr/local/cuda/lib64/liblapack_static.a', needed by 'Open3D/Open3D'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target '/usr/local/cuda/lib64/liblapack_static.a', needed by 'bin/ManuallyCropGeometry'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target '/usr/local/cuda/lib64/liblapack_static.a', needed by 'bin/ConvertPointCloud'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target '/usr/local/cuda/lib64/liblapack_static.a', needed by 'OfflineReconstruction/OfflineReconstruction'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target '/usr/local/cuda/lib64/liblapack_static.a', needed by 'bin/examples/CameraPoseTrajectory'.  Stop.
make[2]: *** Waiting for unfinished jobs....
```


### Open3D, Python and System information

```markdown
- Operating system: Ubuntu 22.04
- Python version: Python 3.10.6
- Open3D version: latest
- System architecture: x86
- Is this a remote workstation?: no
- How did you install Open3D?: build from source
- Compiler version (if built from source): gcc 11.3
```


### Additional information

Looking at the CUDA libraries, it appears that CUDA 12 does not have `liblapack_static.a`, and it now instead has `libcusolver_lapack_static.a`.

--- a/3rdparty/find_dependencies.cmake
+++ b/3rdparty/find_dependencies.cmake
@@ -1686,14 +1686,27 @@ if(BUILD_CUDA_MODULE)
         # open3d_find_package_3rdparty_library, but we have to insert
         # liblapack_static.a in the middle of the targets.
         add_library(3rdparty_cublas INTERFACE)
-        target_link_libraries(3rdparty_cublas INTERFACE
-            CUDA::cusolver_static
-            ${CUDAToolkit_LIBRARY_DIR}/libcusolver_lapack_static.a
-            CUDA::cusparse_static
-            CUDA::cublas_static
-            CUDA::cublasLt_static
-            CUDA::culibos
-        )
+        if(CUDAToolkit_VERSION VERSION_LESS "12.0")
+            target_link_libraries(3rdparty_cublas INTERFACE
+                CUDA::cusolver_static
+                ${CUDAToolkit_LIBRARY_DIR}/liblapack_static.a
+                CUDA::cusparse_static
+                CUDA::cublas_static
+                CUDA::cublasLt_static
+                CUDA::culibos
+            )
+        else()
+            # In CUDA12.0 the liblapack_static.a is deprecated and removed.
+            # Use the libcusolver_lapack_static.a instead.
+            target_link_libraries(3rdparty_cublas INTERFACE
+                CUDA::cusolver_static
+                ${CUDAToolkit_LIBRARY_DIR}/libcusolver_lapack_static.a
+                CUDA::cusparse_static
+                CUDA::cublas_static
+                CUDA::cublasLt_static
+                CUDA::culibos
+            )
+        endif()
         if(NOT BUILD_SHARED_LIBS)
             # Listed in ${CMAKE_INSTALL_PREFIX}/lib/cmake/Open3D/Open3DTargets.cmake.
             install(TARGETS 3rdparty_cublas EXPORT Open3DTargets)

--- a/3rdparty/find_dependencies.cmake
+++ b/3rdparty/find_dependencies.cmake
@@ -1688,7 +1688,7 @@ if(BUILD_CUDA_MODULE)
         add_library(3rdparty_cublas INTERFACE)
         target_link_libraries(3rdparty_cublas INTERFACE
             CUDA::cusolver_static
-            ${CUDAToolkit_LIBRARY_DIR}/liblapack_static.a
+            ${CUDAToolkit_LIBRARY_DIR}/libcusolver_lapack_static.a
             CUDA::cusparse_static
             CUDA::cublas_static
             CUDA::cublasLt_static

--- a/cpp/open3d/core/hashmap/CUDA/CUDAHashBackendBuffer.cu
+++ b/cpp/open3d/core/hashmap/CUDA/CUDAHashBackendBuffer.cu
@@ -24,8 +24,8 @@
 // IN THE SOFTWARE.
 // ----------------------------------------------------------------------------
 
-#include <thrust/sequence.h>
 #include <thrust/execution_policy.h>
+#include <thrust/sequence.h>
 
 #include "open3d/core/CUDAUtils.h"
 #include "open3d/core/hashmap/HashBackendBuffer.h"

--- a/cpp/open3d/core/hashmap/CUDA/CUDAHashBackendBuffer.cu
+++ b/cpp/open3d/core/hashmap/CUDA/CUDAHashBackendBuffer.cu
@@ -25,6 +25,7 @@
 // ----------------------------------------------------------------------------
 
 #include <thrust/sequence.h>
+#include <thrust/execution_policy.h>
 
 #include "open3d/core/CUDAUtils.h"
 #include "open3d/core/hashmap/HashBackendBuffer.h"

